ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.cpp

```diff
@@ -0,0 +1,235 @@
+#include "MockVnodeAndMount.hpp"
+#include "../PrjFSKext/VnodeUtilities.hpp"
+#include "KextMockUtilities.hpp"
+#include <unordered_map>
+#include <sys/errno.h>
+
+using std::string;
+
+typedef std::unordered_map<string, VnodeWeakPointer> PathToVnodeMap;
+typedef std::unordered_map<vnode_t, VnodeWeakPointer> WeakVnodeMap;
+
+static PathToVnodeMap s_vnodesByPath;
+static WeakVnodeMap s_allVnodes;
+
+MountPointer mount::Create(const char* fileSystemTypeName, fsid_t fsid, uint64_t initialInode)
+{
+    MountPointer result(new mount{});
+    assert(strlen(fileSystemTypeName) + 1 < sizeof(result->statfs.f_fstypename));
+    result->statfs.f_fsid = fsid;
+    result->nextInode = initialInode;
+    strlcpy(result->statfs.f_fstypename, fileSystemTypeName, sizeof(result->statfs.f_fstypename));
+    return result;
+}
+
+
+vnode::vnode(const MountPointer& mount) :
+    mountPoint(mount),
+    name(nullptr),
+    inode(mount->nextInode++)
+{
+}
+
+vnode::~vnode()
+{
+    assert(this->ioCount == 0);
+}
+
+VnodePointer vnode::Create(const MountPointer& mount, const char* path, vtype vnodeType)
+{
+    VnodePointer result(new vnode(mount));
+    s_allVnodes.insert(make_pair(result.get(), VnodeWeakPointer(result)));
+    result->weakSelfPointer = result;
+    result->SetPath(path);
+    result->type = vnodeType;
+    return result;
+}
+
+VnodePointer vnode::Create(const MountPointer& mount, const char* path, vtype vnodeType, uint64_t inode)
+{
+    VnodePointer result = Create(mount, path, vnodeType);
+    result->inode = inode;
+    return result;
+}
+
+void vnode::StartRecycling()
+{
+    s_vnodesByPath.erase(this->path);
+    this->path.clear();
+    this->type = VBAD;
+    this->vid++;
+    this->isRecycling = true;
+}
+
+void vnode::SetGetPathError(errno_t error)
+{
+    this->getPathError = error;
+}
+
+void vnode::SetPath(const string& path)
+{
+    s_vnodesByPath.erase(this->path);
+
+    this->path = path;
+    size_t lastSlash = this->path.rfind('/');
+    if (lastSlash == string::npos)
+    {
+        this->name = this->path.c_str();
+    }
+    else
+    {
+        this->name = this->path.c_str() + lastSlash + 1;
+    }
+    
+    assert(!s_vnodesByPath[path].lock());
```

Comment:
A comment here would be useful as well.  As best I can tell this `assert` is validating that this is the first `vnode` being recorded in `s_vnodesByPath` for this path, is that correct?

Reply:
Almost. `s_vnodesByPath[path]` is a reference to a weak pointer; this should either be null (no entry for this path in the map) or fail to be turned into a strong pointer because any previous vnode for that path should have hit a reference count of 0.

So we specifically don't want to check `s_vnodesByPath.find(path) == s_vnodesByPath.end()` because that would incorrectly fail where a vnode used to exist for that path but simply hit a refcount of 0, but the weak pointer is still there.

Reply:
Thanks!

Is it more accurate to say that the assert is to validate that there's not already an active vnode for the specified path?